src/openfermion/utils/_operator_utils.py

```diff
@@ -846,3 +848,97 @@ def normal_ordered(operator, hbar=1.):
+    significantly for some Hamiltonians in the context of
+    VQE (see section V(A) in the supplementary material of
+    https://arxiv.org/pdf/1704.05018v2.pdf). The more general problem
+    of grouping operators into commutitative groups is discussed in
+    section IV (B2) of https://arxiv.org/pdf/1509.04279v1.pdf. The
+    original input operator is the union of all output sub-operators,
+    and all sub-operators are disjoint (do not share any terms).
+
+    Args:
+        operator (QubitOperator): the operator that will be split into
+            sub-operators (tensor product basis sets).
+        randomize (bool): default True. If True, each term in
+            operator will be assigned at random to a compatible
+            sub-operator (or will become a new sub-operator if no compatible
+            sub-operator exists). If False, each term in the operator
+            will be assigned to the compatible sub-operator that was
+            created first (or will become a new sub-operator if no
+            compatible sub-operator exists). A sub-operator is compatible
+            if there exists a tensor product basis in which the term,
+            and all terms already in the sub-operator, are diagonal.
+        seed (int): default None. If None, the random number generator
+            is not re-initialised. Otherwise, the random number generator
+            is re-initialised with the seed.
+
+    Returns:
+        sub_operators (dict): a dictionary where each key defines a
+            tensor product basis, and each corresponding value is a
+            QubitOperator with terms that are all diagonal in
+            that basis.
+            **key** (tuple of tuples): Each key is a term, which defines
+                a tensor product basis. A term is a product of individual
+                factors; each factor is represented by a tuple of the form
+                (`index`, `action`), and these tuples are collected into a
+                larger tuple which represents the term as the product of
+                its factors. `action` is from the set {'X', 'Y', 'Z'} and
+                `index` is a non-negative integer corresponding to the
+                index of a qubit.
+            **value** (QubitOperator): A QubitOperator with terms that are
+                diagonal in the basis defined by the key it is stored in.
+
+    Raises:
+       TypeError: Operator of invalid type.
+    """
+
+    if not isinstance(operator, QubitOperator):
+        raise TypeError('Can only split QubitOperator into tensor product'
+                        ' basis sets. {} is not supported.'.format(
+                            type(operator).__name__))
+
+    sub_operators = OrderedDict()
+    for term, coefficient in operator.terms.items():
+        bases = sub_operators.keys()
+        if randomize:
+            bases = list(bases)
+            if seed:
+                random.seed(seed)
+            random.shuffle(bases)
```

Comment:
I know we haven't been consistent with this in other parts of the code base, but it's better not to have code that changes the global random state. Instead, seed a [numpy.RandomState](https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.RandomState.html#numpy.random.RandomState) and use that to randomize the basis.